code/game/machinery/firealarm.dm

```diff
@@ -326,3 +326,109 @@
 	if (!party_overlay)
 		party_overlay = iconstate2appearance('icons/turf/areas.dmi', "party")
 	A.add_overlay(party_overlay)
+
+/obj/machinery/sprinkler
+	name = "foam dispenser"
+	desc = "An automated sprinkler cabable of detecting fire and spray coolant."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "sprinkler0"
+	layer = PROJECTILE_HIT_THRESHHOLD_LAYER
+	plane = FLOOR_PLANE
+	max_integrity = 100
+	armor = list("melee" = 50, "bullet" = 20, "laser" = 20, "energy" = 20, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 80, "acid" = 30)
+	var/obj/item/reagent_containers/glass/beaker/internal
+	var/refill_loaded = FALSE
+	req_access = list(ACCESS_ATMOSPHERICS)
+	var/cooldown = 0
+
+/obj/machinery/sprinkler/Initialize()
+	. = ..()
+	internal = new /obj/item/reagent_containers/glass/beaker(src)
+
+/obj/machinery/sprinkler/interact(mob/user)
+	if(!allowed(user))
+		to_chat(user, "<span class='danger'>Access denied.</span>")
+		return
+	dispense()
+
+/obj/machinery/sprinkler/fire_act()
+	.=..()
+	if(refill_loaded)
+		dispense()
+
+/obj/machinery/sprinkler/proc/dispense()
+	if(cooldown < world.time && refill_loaded)
+		var/datum/effect_system/smoke_spread/chem/smoke_machine/smoke = new()
+		smoke.set_up(internal.reagents, 5, 8,  get_turf(src))
+		smoke.start()
+		cooldown = world.time + 200
+		if(!internal.reagents)
+			refill_loaded = FALSE
+			update_icon()
+
+/obj/machinery/sprinkler/update_icon()
+	. = ..()
+	if(refill_loaded)
+		icon_state = "sprinkler1"
+		return
+	icon_state = "sprinkler0"
+
+/obj/item/sprinkler_refill
+	name = "sprinkler refill (water)"
+	desc = "Apply this refill if the sprinkler light is red."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "sprinkler_refill_water"
+	item_state = "fire_extinguisher"
+	w_class = WEIGHT_CLASS_SMALL
+	var/obj/item/reagent_containers/glass/beaker/refill
+	var/used = FALSE
+	var/chemtype = /datum/reagent/water
+
+/obj/item/sprinkler_refill/Initialize()
+	. = ..()
+	refill = new /obj/item/reagent_containers/glass/beaker(src)
+	refill.reagents.add_reagent(chemtype, 50)
+
+/obj/item/sprinkler_refill/examine(mob/user)
+	. = ..()
+	if(used)
+		. += "It has been used up, better trow it in the trash."
+
+/obj/item/sprinkler_refill/update_icon()
+	. = ..()
+	cut_overlays()
+	if(used)
+		add_overlay("refill_empty")
+		return
+	add_overlay("refill_[max(round(refill.reagents.total_volume/10),1)]")

```

Comment:
Does this even work? I usually use mutable appearance for overlays.

Reply:
As long as the overlay is one of the `icon_state`s, this should work, I think. `add_overlay` defaults to the atom's icon, I believe; of course, this may require testing.

Reply:
no

Reply:
do the overlays need to be called myiconstate_something to work?

Reply:
Do this and also add an overlay called refill_0 or replace the round with a ceiling.
```
/obj/item/sprinkler_refill/update_icon()
	. = ..()
	cut_overlays()
	add_overlay(mutable_appearance(icon,"refill_[used ? "empty" : max(round(refill.reagents.total_volume/10),1)]"))

Reply:
add_overlay() accepts a text string as an icon state. See overlays in the DM reference